File: Problem2.py
#Time Complexity: O(N)
#Space Complexity: O(N)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        
        self.extra = 0
        self.moves = 0
        self.helper(root)
        return self.moves
    def helper(self,root):
        if not root:
            return 0
        self.extra = root.val - 1
        self.extra += self.helper(root.left)
        self.extra += self.helper(root.right)
        self.moves += abs(self.extra)    
        return self.extra
        

File: problem2.cpp
/*
// Time Complexity : O(N) number of nodes
// Space Complexity : O(N) number of nodes
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
*/

#include<math.h>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left{};
    TreeNode* right{};
    TreeNode():val(0),left(nullptr),right(nullptr){}
    TreeNode(int x):val(x),left(nullptr),right(nullptr){}
    TreeNode(int x, TreeNode* left,TreeNode* right):val(x),left(left),right(right){}
};

class Solution {
public:
    int moves{};
    int dfs(TreeNode* node){
        if(node == NULL) return 0;
        int root_coins{},left_coins{},right_coins{},total_coins{};
        root_coins = node->val - 1;
        left_coins = dfs(node->left);
        right_coins = dfs(node->right);
        total_coins = root_coins + left_coins + right_coins;
        moves+=abs(total_coins);
        return total_coins;
        
    }
    int distributeCoins(TreeNode* root) {
        moves+=abs(dfs(root));
        return moves;
    }
};